### PR TITLE
Remove custom get and custom collections

### PR DESCRIPTION
**Pull request by rklaehn**
## Description

As discussed, this removes the custom get request. Fixes https://github.com/n0-computer/iroh/issues/1554

A custom get request is going to be replaced by the following:

- create a connection under another ALPN (not the iroh-bytes one)
- do whatever custom exchange needed to figure out what is to be done
- use the existing quinn stream for a bao sync

Also removes the custom collection parsing. Collections are now just sequences of blake3 hashes. Fixes https://github.com/n0-computer/iroh/issues/1553

Custom collections are going to be replaced by two things: For the purpose of the network protocol, they are going to be replaced with an exchange under a custom ALPN as described above. For the purpose of GC they are going to be replaced by a way to enumerate live nodes.

## Notes & open questions

How detailed will the examples have to be? It is hard to think of a small example that does not seem arbitrary, but a big example would be a lot of work.

Here is a possible big example: We define a protocol that allows exchanging arbitrary cids, with arbitrary hash fn. We could call it swapbits. We could even implement a more complex protocol to exchange dags, we could call it syncgraph.

Then we implement a node that supports this custom protocol and does the actual exchange using bao.

But as you can imagine this would be a rather large example.

## Change checklist

- [x] Self-review.
- [x] Documentation updates if relevant.
- [x] Tests if relevant.

### CODE REVIEW

**Comment by dignifiedquire** [comment]
> How detailed will the examples have to be?

Having an example that shows how sth like https://github.com/n0-computer/appa could be done would be great (could also just be a PR there, updating it accordingly)

**Comment by rklaehn** [comment]
> > How detailed will the examples have to be?
> 
> Having an example that shows how sth like https://github.com/n0-computer/appa could be done would be great (could also just be a PR there, updating it accordingly)

Yes, that seems reasonable.

File: iroh-bytes/src/protocol.rs [hunk 1]
```diff
@@ -432,40 +432,25 @@ impl Display for RequestToken {
 pub enum Request {
     /// A get request for a blob or collection
     Get(GetRequest),
-    /// A get request that allows the receiver to create a collection
-    CustomGet(CustomGetRequest),
```
[the remark this answers is not in the document]
Comment (rklaehn) [reply iroh-bytes/src/protocol.rs:436]: I want to add a request to find out what the other side has very soon. See https://github.com/n0-computer/iroh/pull/1413 . This should be an integral part of the protocol, not some Custom... thing.

That is why I want to keep this an enum.

File: iroh-bytes/src/collection.rs [hunk 2]
```diff
@@ -49,19 +38,25 @@ impl IntoIterator for LinkSeq {
...
-    fn next(&mut self) -> LocalBoxFuture<'_, anyhow::Result<Option<Hash>>> {
-        futures::future::ok(self.0.pop_front()).boxed_local()
+/// Iterator over the hashes in a [`LinkSeq`].
+#[derive(Debug, Clone)]
+pub struct LinkSeqStream(LinkSeq);
+
+impl LinkSeqStream {
+    /// Get the next hash in the sequence.
+    #[allow(clippy::should_implement_trait)]
+    pub fn next(&mut self) -> impl Future<Output = anyhow::Result<Option<Hash>>> + '_ {
```
Comment (dignifiedquire) [inline iroh-bytes/src/collection.rs:48]: why not use an `async fn`?

**Comment by rklaehn** [comment]
The last commit is a big rename to separate the concept of a sequence of hashes in iroh-bytes (and gc) from the higher level concept of collections that only exists in iroh itself.

Since collections now do not influence the behaviour of iroh-bytes, and it's really not that much code, I think it might be a good idea to get rid of the `iroh-collections` feature flag in iroh.

Also, not sure how everybody feels about it, but `mem-db` is a significant amount of code but does not have any dependencyes. So how about we remove that feature flag as well?

I think `flat-db` is a FF we might want to keep, but that also does not have any extra deps.

**Review by dignifiedquire — APPROVED** [review]
nice

**Pull request merged**